### PR TITLE
add tags

### PR DESCRIPTION
**Pull request by pavlozhuravlov**

### CODE REVIEW

**Comment by rilian** [comment]
please attach screenshots from both site and email with tags

**Comment by pavlozhuravlov** [comment]
Field from link form
image

Links list
image

Email
image

**Comment by rilian** [comment]
ok, but plz make tag a link to a public search of Links with this tag

File: Gemfile [hunk 1]
```diff
@@ -18,6 +19,7 @@ gem 'therubyracer', platforms: :ruby
 gem 'jquery-rails'
 gem 'slim-rails'
 gem 'sprockets-rails', '<3.0' # 3.0 has breaking changes
+gem "select2-rails"
```
Comment (sudodoki) [inline Gemfile]: why double quote instead single?

File: app/assets/javascripts/tags.js [hunk 2]
```diff
@@ -0,0 +1,7 @@
+$( document ).ready(function() {
+  $('.select-tags').select2({
+         tags: true,
+         multiple: true,
+         tokenSeparators: [",", " "]
+  });
```
Comment (tilast) [inline app/assets/javascripts/tags.js:6]: plz prefer using 2 space over the tabs

Comment (sudodoki) [reply app/assets/javascripts/tags.js:6]: there's [editorconfig](http://editorconfig.org/) in repo - please enable this in your editor

**Comment by tilast** [comment]
generally nice

File: app/assets/javascripts/tags.js [hunk 3]
```diff
@@ -0,0 +1,7 @@
+$( document ).ready(function() {
```
Comment (sudodoki) [inline app/assets/javascripts/tags.js]: no need for spaces here

**Review by tilast — APPROVED** [review]
ok for me but let's wait for @rilian review

File: app/controllers/links_controller.rb [hunk 4]
```diff
@@ -3,7 +3,12 @@ class LinksController < ApplicationController
 
   def index
     @q = Link.ransack(params[:q])
-    @links = @q.result.page(params[:page]).order('issue_id IS NOT NULL').order(id: :desc)
+    @links =
+      if params[:tag].present?
+        Link.tagged_with(params[:tag])
+      else
+        @q.result
+      end.page(params[:page]).order('issue_id IS NOT NULL').order(id: :desc)
```
Comment (rilian) [inline app/controllers/links_controller.rb]: use variables instead of chaining to `end`

File: app/views/issue_mailer/notify_subscribers.slim [hunk 5]
```diff
@@ -12,6 +12,8 @@ body topmargin="0" leftmargin="0" marginheight="0" marginwidth="0" style="-webki
       - @issue.links.each do |link|
         li
           a href="#{link.url}" target="_blank" style="font-size: 18px; color: #4183c4; text-decoration: none;" #{link.title}
+          - link.tag_list.each do |tag|
+            span style="display: inline-block; border: 1px solid black; border-radius: 5px; min-width: 40px; margin: 0 10px; padding: 0 5px; text-align: center;" #{tag}
```
Comment (rilian) [inline app/views/issue_mailer/notify_subscribers.slim]: no link here

**Review by rilian — CHANGES_REQUESTED** [review]
i am not sure links search is working for guests

**Review by rilian — APPROVED** [review]

**Comment by rilian** [comment]
not approved yet. testing github UI :)

**Review by rilian — CHANGES_REQUESTED** [review]
plz review comments above

File: app/services/link_filter.rb [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    get_links
+  end
+
+private
+
+  def get_links
+    if @authenticated
+      search_result =
+        if @params[:tag].present?
+          Link.tagged_with(@params[:tag])
```
Comment (tilast) [inline app/services/link_filter.rb:17]: i think we should also add a constraint for the links to check issue id
we should not show links without the issue

Comment (rilian) [reply app/services/link_filter.rb:17]: also links to unpublished issues should not be searchable

Comment (tilast) [reply app/services/link_filter.rb:17]: oh, I meant this

File: app/views/links/search_by_tag.slim [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+                  a href="#{links_path(tag: tag)}" #{tag}
+              br
+              a href="#{link.url}" target="_blank" rel="nofollow" #{link.url}
+              - if link.time_to_read.present? && link.time_to_read > 0
+                span
+                |  (#{link.time_to_read} mins reading)
+              br
+              = link.desc.strip.gsub("\n", '<br/>').html_safe
+              - if link.manager
+              br
```
Comment (tilast) [inline app/views/links/search_by_tag.slim]: it will not be under the `if` - you should add a space here

File: app/views/links/search_by_tag.slim [hunk 8]
```diff
@@ -0,0 +1,32 @@
+.columns
+  .column class='three-fourths'
```
Comment (tilast) [inline app/views/links/search_by_tag.slim]: just

``` slim
  .column.three-fourths
```

**Review by tilast — APPROVED** [review]
looks ok for me but just fix two comments

**Comment by rilian** [comment]
@tilast plz merge once comments will be fixed

**Comment by tilast** [comment]
@pavel-zhuravlyov nice to merge

**Pull request merged**